pkg/backoff/backoff_internal_test.go

```diff
@@ -0,0 +1,85 @@
+package backoff
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/wakatime/wakatime-cli/pkg/config"
+
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpdateBackoffSettings(t *testing.T) {
+	v := viper.New()
+
+	tmpFile, err := ioutil.TempFile(os.TempDir(), "wakatime")
+	require.NoError(t, err)
+
+	defer os.Remove(tmpFile.Name())
+
+	v.Set("config", tmpFile.Name())
+
+	err = updateBackoffSettings(v, 2, nil)
+	require.NoError(t, err)
+
+	ini, err := config.NewIniWriter(v, func(vp *viper.Viper) (string, error) {
+		assert.Equal(t, v, vp)
+		return tmpFile.Name(), nil
+	})
+	require.NoError(t, err)
+
+	assert.NotEmpty(t, ini.File.Section("internal").Key("backoff_at").String())
+	assert.Equal(t, "2", ini.File.Section("internal").Key("backoff_retries").String())
+}
+
+func TestUpdateBackoffSettings_Reset(t *testing.T) {
+	v := viper.New()
+
+	tmpFile, err := ioutil.TempFile(os.TempDir(), "wakatime")
+	require.NoError(t, err)
+
+	defer os.Remove(tmpFile.Name())
+
+	v.Set("config", tmpFile.Name())
+
+	at := time.Now().Add(time.Minute*(resetAfter*-1) + time.Second*-1)
+
+	err = updateBackoffSettings(v, 2, &at)
```

Comment:
What is the advantage of passing `*time.Time` instead of `time.Time` into `updateBackoffSettings`?

Reply:
The key `backoff_at` can be a null value when it's not present in the config file or parsing failed. Since it can be null, a pointer is necessary to the logic for backoff to work. The possible scenarios are:
* not in backoff, api returns success, then nothing changes.
* not in backoff, api returns an error, then it updates (adds) backoff settings. if BackoffAt is a null value it stamps a new time.
* in backoff, api returns an error, then it updates backoff settings. BackoffAt is **not** a null value.
* in backoff, api returns an error, limit time during backoff reached, then it resets backoff settings to retries 1 and stamp a new time. BackoffAt is **not** a null value.